garak/configurable.py

```diff
@@ -0,0 +1,82 @@
+import logging
+import inspect
+from garak import _config
+from garak import _plugins
+
+
+class Configurable:
+    def _load_config(self, config_root=_config):
+        if hasattr(self, "loaded"):
+            return  # only load once, this will ensure the config is not rerun for extending classes
+        local_root = (
+            config_root.plugins if hasattr(config_root, "plugins") else config_root
+        )
+        namespace_parts = self.__module__.split(".")
+        # last part is the namespace, second to last is the plugin type
+        # this will support something like:
+        # plugins['detectors'][x]['generators']['rest']['RestGenerator']
+        # plugins['detectors'][x]['generators']['rest']
+        # plugins['probes'][y]['generators']['rest']['RestGenerator']
+        if len(namespace_parts) > 2:
+            # example class expected garak.generators.huggingface.Pipeline
+            # spec_type = generators
+            # namespace = huggingface
+            # classname = Pipeline
+            # current expected spec_type values are `_plugins.PLUGIN_TYPES`
+            spec_type = namespace_parts[-2]
+            namespace = namespace_parts[-1]
+            classname = self.__class__.__name__
+            plugins_config = {}
+            if isinstance(local_root, dict) and spec_type in local_root:
+                plugins_config = local_root[spec_type]
+            elif hasattr(local_root, spec_type):
+                plugins_config = getattr(local_root, spec_type)
+            if namespace in plugins_config:
+                # example values:
+                # generators: `nim/openai/huggingface`
+                # probes: `dan/gcg/xss/tap/promptinject`
+                attributes = plugins_config[namespace]
+                namespaced_klass = f"{namespace}.{classname}"
+                self._apply_config(attributes)
+                if classname in attributes:
+                    self._apply_config(attributes[classname])
+                elif namespaced_klass in plugins_config:
+                    # for compatibility remove after
+                    logging.warning(
+                        f"Deprecated configuration key found: {namespaced_klass}"
+                    )
+                    self._apply_config(plugins_config[namespaced_klass])
+        self._apply_missing_instance_defaults()
+        self.loaded = True
+
+    def _apply_config(self, config):
+        classname = self.__class__.__name__
+        init_params = inspect.signature(self.__init__).parameters
+        for k, v in config.items():
+            if k in _plugins.PLUGIN_TYPES or k == classname:
+                # skip entries for more qualified items or any plugin type
+                # should this be coupled to `_plugins`?
+                continue
+            if hasattr(self, "_supported_params") and k not in self._supported_params:
```

Comment:
I like the idea 👍

Reply:
```
>>> val = False
>>> if val and val.does_not_exists:
...     print("Still looked")
... else:
...     print("did not check second condition")
...
did not check second condition
```

also if set to `None`:
```
>>> val = None
>>> if val and val.does_not_exists:
...     print("Still looked")
... else:
...     print("did not check second condition")
...
did not check second condition
```

By defining the class variable as `None` this could be updated to:
```suggestion
            if self._supported_params and k not in self._supported_params:
```

Or still default to None to be sure it exists maybe be explicit that the expectation is that a list is provided:
```suggestion
            if isinstance(self._supported_params, list) and k not in self._supported_params:
```

Reply:
Note this conditional causes a value to skip being propagated if the class provides an allow list. Removing the first condition and supplying a default empty list would reject all values for any class that does not provide an allow list.

Requiring an allow list is a more secure pattern but also more restrictive than I suspect is needed at this time.

Reply:
We do need to keep the dev ux reasonably pleasant. I'm thinking a bit about tests and logging to go along with this feature. I guess if one were to set `_supported_params` to an empty list by default, a test could be added to check for this parameter in plugin class defs. Thinking about potentially unintuitive failure modes, I think the things that could go wrong are:
* A value in a given config is not applied - e.g. because a `_config` key is not in `_supported_params` - this should be logged
* A value in the plugin's default config is not applied, because it's not allowed by `_supported_params` - can this happen? - I think this should be a test failure
* A value a plugin expects, is not assigned - e.g. because a key in `_supported_params` doesn't correspond with the plugin's default config's key - this should be logged and perhaps also a test failure